Using ObjectParser in UpdateRequest

liketic opened the pull request:
Reject unknown field and switch ```fromXContent``` to ```ObjectParser```. 

Closes #28740

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Pinging @elastic/es-distributed

nik9000 commented on server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java:
```diff
@@ -55,6 +57,58 @@
+                    fields = Collections.singletonList(parser.text());
+                }
+                return fields;
+            }, FIELDS_FIELD, ObjectParser.ValueType.STRING_ARRAY);
```
I *think* you can use `PARSER.declareStringArray(UpdateRequest::fields, FIELDS_FIELD)`. It doens't reject non-array things but I think that is fine. Another choice would be to simply pitch this. It has been deprecated for a year and a half (since 1764ec56b38) and we are only going to merge this to the master branch anyway so it is going to be released during a major version so we can make breaking changes like dropping a deprecated field.

liketic commented:
@nik9000 Thank you so much! I updated the PR and hope that's what we need. And added a note in migration/api , not sure if that's appropriate.  And another problem is I don't know how to avoid the failure when executing task ```:qa:mixed-cluster:v6.3.0-SNAPSHOT#mixedClusterTestRunner```, because I removed ```fields = in.readOptionalStringArray();``` from ```UpdateRequest.readFrom``` which caused a bwc issue. Thanks for your help!

nik9000 commented on server/src/test/java/org/elasticsearch/update/UpdateIT.java:
```diff
@@ -225,7 +225,7 @@ public void testUpsertDoc() throws Exception {
                 .setDoc(XContentFactory.jsonBuilder().startObject().field("bar", "baz").endObject())
                 .setDocAsUpsert(true)
-                .setFields("_source")
+                .setFetchSource(true)
```
Hmmm. I *think* this is true by default. Can you check what happens if you drop the `.setFetchSource(true)` part?

liketic replied:
By default the field ```fetchSource``` is null in the class UpdateRequest:
```
    private FetchSourceContext fetchSourceContext;
```. 
If we drop this, we'll get a AssertionError.

nik9000 replied:
I see it. I was tired when I was reading it. You are right to do what you did.

nik9000 commented on server/src/test/java/org/elasticsearch/update/UpdateIT.java:
```diff
@@ -293,12 +293,8 @@ public void testUpdate() throws Exception {
-        } catch (DocumentMissingException e) {
-            // all is well
-        }
+        expectThrows(DocumentMissingException.class,
```
:heart:

nik9000 replied:
Usually I do
```
Exception e = expectThrows(DocumentMissingException.class, () ->
    the thing);
assertEquals("whatever", e.getMessage());
```

Just to make sure it was the *right* exception. In this case there is only one kind of message for `DocumentMissingException` so it is probably safe to do what you are doing, but it still might be nice anyway.

nik9000 commented:
And thanks so much for doing this!

@jimczi, I believe you deprecated `fields` in updates. Are you ok with removing for 7.0 like I've suggested here? I just realized I might want to confirm with you.

jimczi commented:
+1 to remove `fields. Thanks @liketic @nik9000 !

liketic commented on server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java:
```diff
@@ -375,21 +370,11 @@ public BulkRequest add(BytesReference data, @Nullable String defaultIndex, @Null
-                                DEPRECATION_LOGGER.deprecated("Deprecated field [fields] used, expected [_source] instead");
-                                List<Object> values = parser.list();
-                                fields = values.toArray(new String[values.size()]);
-                            } else {
```
Thanks! I put the ```throw``` back and add a test.

liketic commented:
Thanks @nik9000 , I pushed 2024751c8be2fb6a649fce616e746778565ff7b6

nik9000 commented on server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java:
```diff
@@ -800,7 +772,6 @@ public void readFrom(StreamInput in) throws IOException {
             doc = new IndexRequest();
             doc.readFrom(in);
         }
-        fields = in.readOptionalStringArray();
```
So removing this line is going to cause problems in mixed clusters. Elasticsearch versions before 7.0.0 will continue to send this array. You have to read it and decide what to do with it. Something like:
```
        if (false == in.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
            String[] fields = in.readOptionalStringArray();
            if (fields != null) {
                throw new IllegalArgumentException("[fields] is no longer supported");
            }
        }
```

is probably fine to be honest.

nik9000 commented on server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java:
```diff
@@ -838,7 +809,6 @@ public void writeTo(StreamOutput out) throws IOException {
             doc.id(id);
             doc.writeTo(out);
         }
-        out.writeOptionalStringArray(fields);
```
You'll want to do something like:
```
if (false == out.getVersion().onOrAfter(Version.V_7_0_0_alpha_1) {
  out.writeOptionalStringArray(null);
}
```

That way 6.x can parse the request. We know there aren't any `fields` anyway because we couldn't have parsed them.

nik9000 commented on docs/reference/migration/migrate_7_0/api.asciidoc:
```diff
@@ -22,3 +22,6 @@ The following parameters starting with underscore have been removed:
 
+
+==== The parameter `fields` deprecated in 6.x has been removed from Bulk request 
+and Update request.
```
Can you also add a note about the change you update request?

I'm not *sure* this is the right file for them but you may as well put them here for now and we'll move them if we decide there is a better spot.

liketic commented:
Thanks @nik9000 That's much better!  I pushed fae79081c531e0829f1597096380a0fce4e039ba

nik9000 commented:
Or maybe there are conflicts. Nuts!

If you are around @liketic, can you merge master into this branch and resolve the conflicts? I can do it if you'd prefer.

liketic commented:
Thanks @nik9000 , the conflict was resolved.

nik9000 commented:
> Thanks @nik9000 , the conflict was resolved.

:heart:

nik9000 commented:
Damn, wrong power words!

@elasticmachine, run all tests!

nik9000 commented:
That did it.

nik9000 commented:
The PR build failed but it looks like that is because this needs master merged into it again. I've done so locally and am re-running the build locally.

nik9000 commented:
OK! I got a passing build locally. I'll merge this when the intake build is green.

nik9000 commented:
Merged to master! Finally! Thanks @liketic! Thanks for sticking with it! I've learned my lesson. Next time I won't ask about dropping a deprecated thing when converting to ObjectParser. That more than doubled the complexity of this. But thanks for finishing it out!

For posterity: I rewrote the commit message but accidentally didn't rewrite the subject line. Oops.
